### PR TITLE
Moving EVD and SVD heavy lifiting into separate methods

### PR DESCRIPTION
**Pull request by izcram**
Small refactoring PR to get to know the project and its processes

### CODE REVIEW

File: src/shogun/preprocessor/PCA.cpp [hunk 1]
```diff
@@ -102,168 +102,176 @@ bool CPCA::init(CFeatures* features)
 
 		// center data
 		Map<MatrixXd> fmatrix(feature_matrix.matrix, num_features, num_vectors);
+
```
Comment (karlnapf) [inline src/shogun/preprocessor/PCA.cpp]: you can put your name to the authors at the beginning of the file
as well as changing the license to BSD (grep for BSD license in the source tree)

File: src/shogun/preprocessor/PCA.cpp [hunk 2]
```diff
@@ -102,168 +102,176 @@ bool CPCA::init(CFeatures* features)
...
+	// right singular vectors form eigenvectors
+	m_transformation_matrix = SGMatrix<float64_t>(num_features, num_dim);
+	Map<MatrixXd> transformMatrix(m_transformation_matrix.matrix, num_features, num_dim);
+	num_old_dim = num_features;
+	transformMatrix = svd.matrixV().block(0, 0, num_features, num_dim);
+
+	if (m_whitening) {
+		for (int32_t i = 0; i < num_dim; i++) {
+			if (CMath::fequals_abs<float64_t>(0.0, eigenValues[i], m_eigenvalue_zero_tolerance)) {
```
Comment (karlnapf) [inline src/shogun/preprocessor/PCA.cpp]: minor style:
We do new lines for {

File: src/shogun/preprocessor/PCA.cpp [hunk 3]
```diff
@@ -102,168 +102,176 @@ bool CPCA::init(CFeatures* features)
...
+void CPCA::init_with_evd(const SGMatrix<float64_t>& feature_matrix, int32_t max_dim_allowed) {
+	int32_t num_vectors = feature_matrix.num_cols;
+	int32_t num_features = feature_matrix.num_rows;
+
+	Map<MatrixXd> fmatrix(feature_matrix.matrix, num_features, num_vectors);
+	Map<VectorXd> eigenValues(m_eigenvalues_vector.vector, max_dim_allowed);
+
+	// covariance matrix
+	MatrixXd cov_mat(num_features, num_features);
+	cov_mat = fmatrix*fmatrix.transpose();
```
Comment (karlnapf) [inline src/shogun/preprocessor/PCA.cpp]: next (small) PR: transpose actually *changes* the matrix memory, which is horrible. adjoint is the method we want to use here

File: src/shogun/preprocessor/PCA.h [hunk 4]
```diff
@@ -225,6 +225,10 @@ class CPCA: public CDimensionReductionPreprocessor
 		 * whitening to tackle numerical issues
 		 */
 		float64_t m_eigenvalue_zero_tolerance;
+
+	private:
+		void init_with_evd(const SGMatrix<float64_t>& feature_matrix, int32_t max_dim_allowed);
```
Comment (karlnapf) [inline src/shogun/preprocessor/PCA.h]: These needs minimal doxygen documentation saying that they alter the state (transformation matrix).
Also potentially could say that EVD is based on an eigendocoposition of the covariance matrix, while SVD does an SVD on the data matrix.

**Comment by karlnapf** [comment]
One final request: Could you squash the commits? (or using git amend and force push to your fork)

**Comment by lisitsyn** [comment]
@karlnapf you can squash commits with a click now. "Squash and merge"

**Pull request merged**